File: py/param_manager/param_manager.py
import os


class ParamManager:
    def __init__(self, dir: str = None):
        # global
        self.x_col: str = "x"
        self.y_col: str = "y"
        # ridge finding
        self.sigma: float = 1.0
        self.threshold: float = 5.0e-4
        self.num_bins_x: int = 1000
        self.num_bins_y: int = 1000
        # lowess smoothing
        self.bandwidth: float = 0.1
        self.polynomial_order: int = 2
        # linearize
        self.num_extra_lines: int = 0

        file = os.path.join(dir, "params.dat")
        self.read_params_from_file(file)

    def read_params_from_file(self, file: str):
        for line in open(file, "r"):
            line = line.strip()
            if line.strip() == "":
                continue
            if line[0] == "#":
                continue
            words = line.split(":")

            param_name = words[0].strip()
            param_value = words[1].strip()

            match param_name:
                case "x_col":
                    self.x_col = param_value
                case "y_col":
                    self.y_col = param_value
                case "sigma":
                    self.sigma = float(param_value)
                case "threshold":
                    self.threshold = float(param_value)
                case "bins_x":
                    self.num_bins_x = int(param_value)
                case "bins_y":
                    self.num_bins_y = int(param_value)
                case "bandwidth":
                    self.bandwidth = float(param_value)
                case "polynomial_order":
                    self.polynomial_order = int(param_value)
                case "num_extra_lines":
                    self.num_extra_lines = int(param_value)
                case _:
                    print(f"Parameter {param_name} not recognized or used.")

        # self.bandwidth: float = 0.1
        # self.polynomial_order: int = int

        # file = os.path.join(dir, "params.dat")
        # self.read_params_from_file(file)


if __name__ == "__main__":
    ParamManager("data/faust_det_60")

File: src/param_manager.rs
use std::path::Path;

pub struct ParamManager {
    x_col: String,
    y_col: String,
    sigma: f64,
    threshold: f64,
    bins_x: u64,
    bins_y: u64,
    bandwidth: f64,
    polynomial_order: u64,
    num_extra_lines: u64,
}

impl Default for ParamManager {
    fn default() -> Self {
        Self {
            // gloabal
            x_col: "x".into(),
            y_col: "y".into(),
            // ridge finding
            bins_x: 1_024,
            bins_y: 1_024,
            sigma: 1.0,
            threshold: 5e-4,
            // lowess
            polynomial_order: 1,
            bandwidth: 0.1,
            // linearization
            num_extra_lines: 2,
        }
    }
}

// probably should not be using anyhow here but whatever tbh
use anyhow::Result;

impl ParamManager {
    pub fn new(dir_path: &str) -> Result<Self> {
        let mut result = Self::default();

        let param_file = Path::new(dir_path).join("params.dat");
        let param_data = std::fs::read_to_string(param_file)?;
        for (idx, line) in param_data.lines().enumerate() {
            if line.trim().starts_with("#") {
                continue;
            }
            if line.trim().is_empty() {
                continue;
            }
            let words = line.split(":").collect::<Vec<_>>();
            if words.len() < 2 {
                println!(
                    "Warning: Line {} : \"{}\" is improperly formatted ",
                    idx, line
                );
            }
            let param_name = words[0].trim();
            let param_value = words[1].trim();
            match param_name {
                "x_col" => result.x_col = param_value.into(),
                "y_col" => result.y_col = param_value.into(),
                "bins_x" => result.bins_x = param_value.parse::<u64>()?,
                "bins_y" => result.bins_y = param_value.parse::<u64>()?,
                "sigma" => result.sigma = param_value.parse::<f64>()?,
                "threshold" => result.threshold = param_value.parse::<f64>()?,
                "bandwidth" => result.bandwidth = param_value.parse::<f64>()?,
                "polynomial_order" => result.polynomial_order = param_value.parse::<u64>()?,
                "num_extra_lines" => result.num_extra_lines = param_value.parse::<u64>()?,
                _ => {
                    println!("Parameter \"{}\" not recognized and was ignored", param_name)
                }
            }
        }

        //for line in f

        Ok(result)
    }

    pub fn x_col(&self) -> &str {
        &self.x_col
    }
    pub fn y_col(&self) -> &str {
        &self.y_col
    }
    pub fn num_bins_x(&self) -> u64 {
        self.bins_x
    }
    pub fn num_bins_y(&self) -> u64 {
        self.bins_y
    }
    pub fn threshold(&self) -> f64 {
        self.threshold
    }
    pub fn sigma(&self) -> f64 {
        self.sigma
    }
    pub fn bandwidth(&self) -> f64 {
        self.bandwidth
    }
    pub fn polynomial_order(&self) -> u64 {
        self.polynomial_order
    }
    pub fn num_extra_lines(&self) -> u64 {
        self.num_extra_lines
    }
}
